src/compiler/factory/nodeChildren.ts

```diff
@@ -1,14 +1,20 @@
-import { Node } from "../_namespaces/ts";
+import {
+    isToken,
+    Node,
+} from "../_namespaces/ts";
 
-const nodeChildren = new WeakMap<Node, Node[] | undefined>();
+const nodeChildren = new WeakMap<Node, readonly Node[] | undefined>();
+
+const emptyChildren: readonly Node[] = Object.freeze([]);

```

Comment:
Why not use `emptyArray`?

Reply:
Confusingly, services defines its own mutable empty array (and none of them actually guard against modification)

Reply:
Oh, this isn't even in services, yeah, emptyArray would be nice. (I used to have a PR that froze it but it didn't really seem to matter)

Reply:
From the description:

> If we want, we can probably just use the internal `emptyArray`, but I don't think we've ever frozen that one.

Happy to switch over to that too, but Jake mentioned that he didn't want to make that change since `getChildren()` always was declared as a mutable array. Freezing this array (for runtime safety) and propagating the change (for static verification) helped validate that it's unlikely we mutate these arrays.

Reply:
Does freezing `emptyArray` affect our perf at all? If not, I'd just do that and use `emptyArray`.

Reply:
I tried in https://github.com/microsoft/TypeScript/pull/55755 but I think that was actually more of a problem with trying to make immutable empty maps/sets. Sent #58240 to try just arrays quick.

Reply:
it sure looks like freezing is bad: https://github.com/microsoft/TypeScript/pull/58240#issuecomment-2064751428